---
title: "MOFA downstream analysis report"
author: 
  - Purva Kulkarni^[Radboud University Medical Center, purva.kulkarni@radboudumc.nl]
  - Casper de Visser^[Radboud University Medical Center, casper.devisser@radboudumc.nl]
  - Anna Niehues^[Radboud University Medical Center, anna.niehues@radboudumc.nl]
output: html_document
params:
  modelFile: MOFAmodel.hdf5
  metaData: phenotype_covariates_data.csv
  mcaCoord: cbcl_mca_coord.csv
  epic_anno: epigenomics_annotation_out.RData
  MOFA_factor_top100_epigenomics_weights_outdir: "MOFA_factor_top100_epigenomics_weights" 
---

## Introduction

This notebook contains the downstream analysis steps and visualizations in the form of plots generated using a MOFA (Multi-omics Factor Analysis) model. 
The text explaining these analyses steps is adapted from "MOFA+: downstream analysis (in R)" by Ricard Argelaguet and Britta Velten (2020). 

## List input parameters

```{r params, warning = FALSE}
print(params)
```

## Load libraries

```{r loadpackages, message = FALSE}
library(ggplot2)
library(MOFA2)
```


## Load the mofa model 
```{r loadmodel}
model <- load_model(params$modelFile)
```

## Read the meta data file

```{r readMetaData}
metaData <- read.csv(params$metaData)
#head(metaData)
```

## Load DNA methylation annotation file
```{r readAnnotation}
load(params$epic_anno)
```

## Read the file resulting from the multiple correspondance analysis

```{r readMCACoord}
cbcl_mca_coord <- read.csv(params$mcaCoord)
#head(cbcl_mca_coord)
```

## Plot data overview

Obtain an overview of the input omics data. It shows how many views (rows) and how many groups (columns) exist, what are their corresponding dimensionalities and how many missing information they have (grey bars).

```{r overview}
plot_data_overview(model)
```


## Add meta data to the model
The metadata is stored as a data.frame object in model@samples_metadata, and it requires at least the column sample

```{r addmetadata}
names(metaData)[names(metaData) == 'X'] <- 'sample' 
metaData_cor <- metaData
metaData <- metaData[metaData$sample %in% colnames(model@data$metabolomics$group1),]
samples_metadata(model) <- metaData
```


## Variance decomposition

This step quantifies the amount of variance explained ($R^2$) by each factor in each data modality.

### Total variance explained per view
```{r varianceexplained}
knitr::kable(head(get_variance_explained(model)$r2_total[[1]]))
```

### Variance explained for every factor in per view
```{r varianceperfactor}
knitr::kable(head(get_variance_explained(model)$r2_per_factor[[1]]))
```

### Plot the variance explained estimates
```{r varianceestimates}
plot_variance_explained(model, x="view", y="factor") +
  # add explained variance values
  geom_text(label = round(c(
    get_variance_explained(model)$r2_per_factor$group1), 5))

plot_variance_explained(model, x="view", y="factor", plot_total = TRUE)[[2]]
```

## Visualisations of factors

The MOFA factors capture the global sources of variability in the data. Mathematically, each factor ordinates cells along a one-dimensional axis centered at zero. The value per se is not important, only the relative positioning of samples matters. Samples with different signs manifest opposite “effects” along the inferred axis of variation, with higher absolute value indicating a stronger effect.

### Plot factors

Individual factors can be plotted and visualized in combination with other factors in the form of scatter plots and colored based on specific condition, for example, in the plot below factors 1 to 3 have been visualized and are colored based on Sex. 

```{r plotfactors}
 plot_factors(model, factors = 1:3, color_by = "Sex")
```

### Plot factor correlations

```{r factorcor}
plot_factor_cor(model) 
```

## Visualisation of feature weights

The weights provide a score for how strong each feature relates to each factor. Features with no association with the factor have values close to zero, while features with strong association with the factor have large absolute values. The sign of the weight indicates the direction of the effect: a positive weight indicates that the feature has higher levels in the cells with positive factor values, and vice versa.

### Plot top weights per factor epigenomics

Display highest loading features
Tables below each plot show CpG locations on the genome 

```{r topweights_factor_epigenomics, results = "asis"}
require(knitr)

dir.create(params$MOFA_factor_top100_epigenomics_weights_outdir)

for (MOFA_factor in 1:10) {
  # plot top10 weights
  plot(plot_top_weights(model, factors = MOFA_factor, view = 1))
  # retrieve weights
  df_weights <- get_weights(
    model, factors = MOFA_factor, view = 1, abs = TRUE, as.data.frame = TRUE)
  row.names(df_weights) <- df_weights$feature
  # sort weights
  df_weights <- df_weights[order(df_weights$value, decreasing = TRUE), ]
  # get EPIC annotations
  columns <- c("chromosome", "position", "gene.symbol", "gene.accession")
  mapped_CpG_sites <- epicanno[rownames(df_weights)[1:100], columns]
  mapped_CpG_sites$factor.weight <- df_weights$value[1:100]
  # save top100 weights to csv
  write.csv(mapped_CpG_sites,
            file.path(params$MOFA_factor_top100_epigenomics_weights_outdir,
                      paste0("factor", MOFA_factor, "_weights.csv")))
  # display top10 weights
  print(knitr::kable(mapped_CpG_sites[1:10, ]))
  cat("\n\n\n")
}

```



### Plot top weights per factor metabolomics

Display only the top features with highest loading
```{r topweights}
lapply(1:10, function(f) {
  plot_top_weights(model, factors = f, view = 2) 
})
```


### Correlation with behavorial data

```{r corrbehavorialdata}
library(corrplot)
factors <- MOFA2::get_factors(model, factors = "all")
rownames(cbcl_mca_coord)<- cbcl_mca_coord[,1]
cbcl_mca_coord<-cbcl_mca_coord[,-1]
factors <- as.data.frame(factors)

common_rows <- intersect(rownames(cbcl_mca_coord), rownames(factors))
cbcl_mca_coord_subset <- cbcl_mca_coord[rownames(cbcl_mca_coord) %in% common_rows,]
colnames(cbcl_mca_coord_subset) <- gsub(
  "Dim.", "Dim", paste0("MCA.", colnames(cbcl_mca_coord_subset)))
factors_subset <- factors[rownames(factors) %in% common_rows,]
colnames(factors_subset) <- gsub("group1", "MOFA", colnames(factors_subset))

correlation_cbcl_MOFA <- cor(
  cbcl_mca_coord_subset, factors_subset, 
  use = "pairwise.complete.obs", method = "spearman") #default: pearson
my_lim <- max(abs(c(min(correlation_cbcl_MOFA), max(correlation_cbcl_MOFA))))

# significance test - FDR-corrected p-values (Benjamini-Hochberg)
sigTest = psych::corr.test(
  cbcl_mca_coord_subset, factors_subset,
  use = "pairwise",
  method = "spearman",
  adjust = "BH")

corrplot(
  correlation_cbcl_MOFA,
  is.corr = FALSE,
  col = COL2("RdBu", 200),
  col.lim = c(-my_lim, my_lim),
  p.mat = sigTest$p.adj
)
```


### Correlation with phenotype covariates

```{r corrcovariates}
rownames(metaData_cor) <- metaData_cor[,1]
metaData_cor <- metaData_cor[,-1]

commons_rows <- intersect(rownames(metaData_cor), rownames(factors))
metaData_subset <- metaData_cor[rownames(metaData_cor) %in% common_rows,]
metaData_subset2 <- metaData_subset[, c('Age', 'Aggression_Tscore', 'Familynumber')]
metaData_subset <- metaData_subset[, c('Age', 'Aggression_Tscore')]

# correlations
correlation_metadata_MOFA <- cor(
  metaData_subset, factors_subset, use = "pairwise.complete.obs",
  method = "spearman") #default: pearson
my_lim <- max(abs(c(min(correlation_metadata_MOFA), 
                    max(correlation_metadata_MOFA))))

# significance test - FDR-corrected p-values (Benjamini-Hochberg)
sigTest = psych::corr.test(
  metaData_subset, factors_subset,
  use = "pairwise",
  method = "spearman",
  adjust = "BH")

# plot
corrplot(
  correlation_metadata_MOFA,
  is.corr = FALSE,
  col = COL2("RdBu", 200),
  col.lim = c(-my_lim, my_lim),
  p.mat = sigTest$p.adj
)
```

### Boxplot Factors vs. Sex
```{r boxplot covariates (Sex)}
# get factor values per male/female
factors_male <- get_factors(
  model, factor = 1:10)$group1[metaData[metaData$Sex == "Male", ]$sample, ]
factors_female <- get_factors(
  model, factor = 1:10)$group1[metaData[metaData$Sex == "Male", ]$sample, ]
# Wilcoxon rank sum test
wilcox_MOFA_sex <- do.call(rbind, lapply(1:10, function(x) {
  data.frame(
    factor = x,
    p.value = wilcox.test(factors_male[, x], factors_female[, x],
                          alternative = "two.sided")$p.value)
}))
wilcox_MOFA_sex$p.adj <- p.adjust(wilcox_MOFA_sex$p.value, method = "BH")
knitr::kable(wilcox_MOFA_sex)

# box plot
plot_factor(
  model,
  factor = 1:10,
  color_by = "Sex",
  dodge = TRUE,
  add_boxplot = TRUE
) +
  facet_wrap(~factor, nrow = 2, scales = "free",
             labeller = labeller(factor = wilcox_MOFA_sex))

```

### Check potential confounding factors

### Boxplot Factors vs. Sick
```{r boxplot covariates (Sick)}
plot_factor(
  model,
  factor = 1:10,
  color_by = "Sick",
  dodge= TRUE,
  add_boxplot = TRUE
) +
  facet_wrap(~factor, nrow = 2, scales = "free")

```

### Boxplot Factors vs. Menstruation
```{r boxplot covariates (Menstruation)}
plot_factor(
  model,
  factor = 1:10,
  color_by = "Menstruation",
  dodge= TRUE,
  add_boxplot = TRUE
) +
  facet_wrap(~factor, nrow = 2, scales = "free")

```

### Boxplot Factors vs. Vitamine use
```{r boxplot covariates (Vitamines)}
plot_factor(
  model,
  factor = 1:10,
  color_by = "Vitamines",
  dodge= TRUE,
  add_boxplot = TRUE
) +
  facet_wrap(~factor, nrow = 2, scales = "free")

```

### GEE model to correct for twin effect
```{r, message = FALSE, warning = FALSE}
library(gee)


#' Function for GEE model
#'
#' @param data Dataframe with pheno metadata
#' @param pheno_feature Feature to select as outcome
#' @return GEE model
geemodel <- function (data, pheno_feature) {

f <- as.formula(
	paste(pheno_feature,
		paste('mofa_factor'),
		sep = " ~ "))

invisible(gee(f, data=data, id=Familynumber, family=gaussian, corstr="exchangeable", maxiter=100, na.action=na.omit, silent=T))
}



#' Function to run the GEE model with mofa factors as predictor for phenotypic feature
#'
#' @param data Dataframe with pheno metadata
#' @param pheno_feature Feature to select as outcome
#' @return end table with estimates and p-values
run_geemodel <- function(data, pheno_feature) {


# Make dataframes for storing statistics
estimates <- matrix(NA, 10, 2)
robustSE <- matrix(NA, 10, 2)
robustZ <- matrix(NA, 10, 2)
pval <- matrix(NA, 10, 2)


# Run GEE by iterating over the mofa factors (1-10)
for (i in seq(1:10)) {
	data$mofa_factor <- factors_subset[,i]
	coeff <- summary(invisible(geemodel(data, pheno_feature)))$coefficients
	
	for (j in 1:2) {
		estimates[i, j] <- coeff[j,1]
		robustSE[i, j] <- coeff[j,4]
		robustZ[i, j] <- coeff[j,5]
		pval[i, j] <- 2*pnorm(-abs(coeff[j,5]))
	}
}


# Make column names for end table
estimates_names <- lapply(rownames(coeff), function(x) paste0("Estimate_", x)) 
robustse_names <- lapply(rownames(coeff), function(x) paste0("RobustSE_", x))
robustz_names <- lapply(rownames(coeff), function(x) paste0("RobustZ_", x))
pval_names <- lapply(rownames(coeff), function(x) paste0("Pval_", x))

gee_table_colnames <- c(estimates_names, robustse_names, robustz_names, pval_names)

# Make rownames for end table
gee_table_rownames <- lapply(seq(1:10), function(x) paste0("Mofa_factor_", x))

# Make final table
statistics_gee <- cbind(estimates, robustSE, robustZ, pval)
statistics_gee <- data.frame(statistics_gee)
rownames(statistics_gee) <- gee_table_rownames
colnames(statistics_gee) <- gee_table_colnames

# Ad columns with adjusted p-values
statistics_gee$P_adjusted_Intercept <- p.adjust(as.numeric(statistics_gee[,7]))
statistics_gee$P_adjusted_Mofafactor <- p.adjust(as.numeric(statistics_gee[,8]))

return(statistics_gee)

}

```

```{r}
str(cbcl_mca_coord_subset)
```

### GEE model with phenotypic covariates as outcome
```{r,  message = FALSE, warning = FALSE, results = 'hide'}

metaData_subset2 <- metaData_subset2[order(metaData_subset2$Familynumber),]

geemodel_age <- run_geemodel(metaData_subset2, 'Age')
geemodel_agg <- run_geemodel(metaData_subset2, 'Aggression_Tscore')
```

### Print GEE models
```{r, echo=FALSE}
knitr::kable(geemodel_age, caption="GEE model with Age as outcome")
knitr::kable(geemodel_agg, caption="GEE model with Aggression T-score as outcome")
```


### GEE model with MCA dimension coordinates as outcome

```{r, message = FALSE, warning = FALSE, results='hide'}

cbcl_mca_coord_subset2 <- cbcl_mca_coord_subset

cbcl_mca_coord_subset2$Familynumber <- metaData_subset2$Familynumber 
geemodel_mca1 <- run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim1')
geemodel_mca2 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim2')
geemodel_mca3 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim3')
geemodel_mca4 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim4')
geemodel_mca5 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim5')
geemodel_mca6 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim6')
geemodel_mca7 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim7')
geemodel_mca8 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim8')
geemodel_mca9 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim9')
geemodel_mca10 <-run_geemodel(cbcl_mca_coord_subset2, 'MCA.Dim10')
```

### Print GEE models
```{r, echo=FALSE}
knitr::kable(geemodel_mca1, caption="GEE model with MCA dimension 1 as outcome")
knitr::kable(geemodel_mca2, caption="GEE model with MCA dimension 2 as outcome")
knitr::kable(geemodel_mca3, caption="GEE model with MCA dimension 3 as outcome")
knitr::kable(geemodel_mca4, caption="GEE model with MCA dimension 4 as outcome")
knitr::kable(geemodel_mca5, caption="GEE model with MCA dimension 5 as outcome")
knitr::kable(geemodel_mca6, caption="GEE model with MCA dimension 6 as outcome")
knitr::kable(geemodel_mca7, caption="GEE model with MCA dimension 7 as outcome")
knitr::kable(geemodel_mca8, caption="GEE model with MCA dimension 8 as outcome")
knitr::kable(geemodel_mca9, caption="GEE model with MCA dimension 9 as outcome")
knitr::kable(geemodel_mca10, caption="GEE model with MCA dimension 10 as outcome")
```




```{r}
sessionInfo()
```


