{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "72bae153",
   "metadata": {},
   "outputs": [],
   "source": [
    "# More information on:\n",
    "# https://github.com/ResearchObject/ro-crate-py; https://about.workflowhub.eu/Workflow-RO-Crate/\n",
    "\n",
    "# Import modules\n",
    "from rocrate.rocrate import ROCrate\n",
    "from rocrate.model.file import File\n",
    "from rocrate.model.computationalworkflow import ComputationalWorkflow\n",
    "from rocrate.model.computerlanguage import ComputerLanguage\n",
    "from rocrate.model.person import Person\n",
    "from rocrate.model.contextentity import ContextEntity\n",
    "from rocrate.model.data_entity import DataEntity\n",
    "\n",
    "# Intialize the RO Crate object\n",
    "crate = ROCrate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "246f25af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get current date, which will be added as publish date\n",
    "\n",
    "from datetime import date\n",
    "today = date.today()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f21f9e",
   "metadata": {},
   "source": [
    "## Main Workflow file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "160691af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the Nextflow workflow as a ComputationalWorkflow, add to crate.\n",
    "\n",
    "MainWorkflow = crate.add(ComputationalWorkflow(crate, \"action.nf\", properties={\n",
    "    \"@id\": \"action.nf\",\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"NTR-ACTION Data-analysis workflow\",\n",
    "    \"dateCreated\": str(today),\n",
    "    \"input\": \"\",\n",
    "    \"description\": \"Use of multi-omics data (Metabolomics + DNA Methylation) to study CBCL data\",\n",
    "    \"output\": \"\",\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"url\": \"https://github.com/Xomics/ACTIONdemonstrator_workflow\",\n",
    "    \"version\": \"1.0.0\"\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f2c0ecf",
   "metadata": {},
   "source": [
    "## Contextual information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ee16785d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define Nextflow as computer language\n",
    "\n",
    "nextflow_id = \"https://w3id.org/workflowhub/workflow-ro-crate#nextflow\"\n",
    "\n",
    "Nextflow = crate.add (ComputerLanguage(crate, nextflow_id, properties={\n",
    "  \"@id\": \"https://w3id.org/workflowhub/workflow-ro-crate#nextflow\",\n",
    "  \"@type\": \"ComputerLanguage\",\n",
    "  \"name\": \"Nextflow\",\n",
    "  \"identifier\": {\n",
    "    \"@id\": \"https://www.nextflow.io/\"\n",
    "  },\n",
    "  \"url\": {\n",
    "    \"@id\": \"https://www.nextflow.io/\"\n",
    "  }\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e6a2fdf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define persons (authors)\n",
    "\n",
    "Anna_Niehues_id = \"https://orcid.org/0000-0002-9839-5439\"\n",
    "Casper_de_Visser_id = \"https://orcid.org/0000-0002-2812-5898\"\n",
    "Fiona_Hagenbeek_id = \"https://orcid.org/0000-0002-8773-0430\"\n",
    "Naama_Karu_id = \"https://orcid.org/0000-0001-8005-0726\"\n",
    "Alida_Kindt_id = \"https://orcid.org/0000-0001-6551-6030\"\n",
    "Purva_Kulkarni_id = \"https://orcid.org/0000-0002-4681-4582\"\n",
    "Rene_Pool_id = \"https://orcid.org/0000-0001-5579-0933\"\n",
    "Dorret_Boomsma_id = \"https://orcid.org/0000-0002-7099-7972\"\n",
    "Jenny_van_Dongen_id = \"https://orcid.org/0000-0003-2063-8741\"\n",
    "Alain_van_Gool_id =  \"https://orcid.org/0000-0003-0010-5286\"\n",
    "PeterBram_t_Hoen_id = \"https://orcid.org/0000-0003-4450-3112\"\n",
    "\n",
    "Anna_Niehues = crate.add(Person(crate, Anna_Niehues_id, properties={\n",
    "    \"name\": \"Anna Niehues\",\n",
    "    \"affiliation\": \"Radboud university medical center\"\n",
    "}))\n",
    "Casper_de_Visser = crate.add(Person(crate, Casper_de_Visser_id, properties={\n",
    "    \"name\": \"Casper de Visser\",\n",
    "    \"affiliation\": \"Radboud university medical center\"\n",
    "}))\n",
    "Fiona_Hagenbeek = crate.add(Person(crate, properties={\n",
    "    \"name\": \"Fiona A. Hagenbeek\",\n",
    "    \"affiliation\": \"Vrije Universiteit Amsterdam\"\n",
    "}))\n",
    "Naama_Karu = crate.add(Person(crate, Naama_Karu_id, properties={\n",
    "    \"name\": \"Naama Karu\",\n",
    "    \"affiliation\": \"Leiden University\"\n",
    "}))\n",
    "Alida_Kindt = crate.add(Person(crate, Alida_Kindt_id, properties={\n",
    "    \"name\": \"Alida S.D. Kindt\",\n",
    "    \"affiliation\": \"Leiden University\"\n",
    "}))\n",
    "Purva_Kulkarni = crate.add(Person(crate, Purva_Kulkarni_id, properties={\n",
    "    \"name\": \"Purva Kulkarni\",\n",
    "    \"affiliation\": \"Radboud university medical center\"\n",
    "}))\n",
    "Rene_Pool = crate.add(Person(crate, Rene_Pool_id, properties={\n",
    "    \"name\": \"Ren√© Pool\",\n",
    "    \"affiliation\": \"Vrije Universiteit Amsterdam\"\n",
    "}))\n",
    "Dorret_Boomsma = crate.add(Person(crate, Dorret_Boomsma_id, properties={\n",
    "    \"name\": \"Dorret I. Boomsma\",\n",
    "    \"affiliation\": \"Vrije Universiteit Amsterdam\"\n",
    "}))\n",
    "Jenny_van_Dongen = crate.add(Person(crate, Jenny_van_Dongen_id, properties={\n",
    "    \"name\": \"Jenny van Dongen\",\n",
    "    \"affiliation\": \"Vrije Universiteit Amsterdam\"\n",
    "}))\n",
    "Alain_van_Gool = crate.add(Person(crate, Alain_van_Gool_id, properties={\n",
    "    \"name\": \"Alain J. van Gool\",\n",
    "    \"affiliation\": \"Radboud university medical center\"\n",
    "}))\n",
    "PeterBram_t_Hoen = crate.add(Person(crate, PeterBram_t_Hoen_id, properties={\n",
    "    \"name\": \"Peter A.C. 't Hoen\",\n",
    "    \"affiliation\": \"Radboud university medical center\"\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ca07ea64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define X-omics organization\n",
    "# TODO: Can another url identifier be used here?\n",
    "\n",
    "x_omics_id = \"https://x-omics.nl/\"\n",
    "\n",
    "x_omics = crate.add(ContextEntity(crate, x_omics_id, properties={\n",
    "    \"@type\": \"Organization\",\n",
    "    \"name\": \"The Netherlands X-omics intiative\",\n",
    "    \"url\": \"https://x-omics.nl/\"  \n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c041ff",
   "metadata": {},
   "source": [
    "## ISA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6dd5f713",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define ISA files\n",
    "\n",
    "# Define data entity IDs\n",
    "investigation_id = \"Synthetic_data/i_investigation.txt\"\n",
    "study_id = \"Synthetic_data/s_study.txt\"\n",
    "assay_epigenomics_id = \"Synthetic_data/a_assay_methylation.txt\"\n",
    "assay_amines_id = \"Synthetic_data/a_assay_metabolomics_amines.txt\"\n",
    "assay_oa_id = \"Synthetic_data/a_assay_metabolomics_OA.txt\"\n",
    "assay_steroids_id = \"Synthetic_data/a_assay_metabolomics_steroids.txt\"\n",
    "\n",
    "\n",
    "\n",
    "investigation = crate.add(File(crate, investigation_id, dest_path=investigation_id, properties={\n",
    "    \"@id\": investigation_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Investigation file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C165216\", #Experiment Metadata\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/OBI_0000066\"}, #investigation\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3687\"} #ISA-TAB\n",
    "}))\n",
    "\n",
    "study = crate.add(File(crate, study_id, dest_path=study_id, properties={\n",
    "    \"@id\": study_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Study file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\":  {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C165216\", #Experiment Metadata\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/NCIT_C63536\"}, #study\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3687\"} #ISA-TAB\n",
    "}))\n",
    "\n",
    "assay_epigenomics = crate.add(File(crate, assay_epigenomics_id, dest_path=assay_epigenomics_id, properties={\n",
    "    \"@id\": assay_epigenomics_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Assay epigenomics file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\":  {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C165216\", #Experiment Metadata\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/OBI_0000070\"}, #assay\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3687\"} #ISA-TAB\n",
    "}))\n",
    "\n",
    "assay_amines = crate.add(File(crate, assay_amines_id, dest_path=assay_amines_id, properties={\n",
    "    \"@id\": assay_amines_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Assay amines file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\":  {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C165216\", #Experiment Metadata\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/OBI_0000070\"}, #assay\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3687\"} #ISA-TAB\n",
    "}))\n",
    "\n",
    "assay_oa = crate.add(File(crate, assay_oa_id, dest_path=assay_oa_id, properties={\n",
    "    \"@id\": assay_oa_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Assay organic acids file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\":  {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C165216\", #Experiment Metadata\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/OBI_0000070\"}, #assay\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3687\"} #ISA-TAB\n",
    "}))\n",
    "\n",
    "assay_steroids = crate.add(File(crate, assay_steroids_id, dest_path=assay_steroids_id, properties={\n",
    "    \"@id\": assay_steroids_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Assay steroids file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\":  {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C165216\", #Experiment Metadata\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/OBI_0000070\"}, #assay\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3687\"} #ISA-TAB\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a72e6713",
   "metadata": {},
   "source": [
    "## Sub-workflows with input/output files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f6effb9",
   "metadata": {},
   "source": [
    "### Analyze missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8a048c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define entity IDs\n",
    "missing_data_heatmap_id = \"modules/heatmap_missingness.nf\"\n",
    "missing_data_script =  \"bin/heatmap_missingness.R\"\n",
    "\n",
    "\n",
    "# Heatmap missing data points\n",
    "missing_data_heatmap = crate.add(ComputationalWorkflow(crate, missing_data_heatmap_id, dest_path=missing_data_heatmap_id, properties={\n",
    "    \"@id\": missing_data_heatmap_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Heatmap missingness\",\n",
    "    \"input\": {\"@id\": \"epigenomics_values\",\n",
    "              \"@id\": \"metabolomics_values\",\n",
    "              \"@id\": \"behavioral_data\",\n",
    "              \"@id\": \"phenotype_covariates\"},  \n",
    "    #\"output\": {} ,\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": missing_data_script},\n",
    "\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C142610\", #Missing Data\n",
    "                       \"@id\": \"http://semanticscience.org/resource/SIO_000449\"} #plot \n",
    "\n",
    "}))\n",
    "\n",
    "\n",
    "missing_data_script = crate.add(File(crate, missing_data_script, dest_path=missing_data_script, properties={\n",
    "    \"@id\": missing_data_script,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Heatmap NA values script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C142610\", #Missing Data\n",
    "                       \"@id\": \"http://semanticscience.org/resource/SIO_000449\"} #plot \n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ce26b7c",
   "metadata": {},
   "source": [
    "### Epigenomics pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "bcc2145b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define entity IDs\n",
    "epi_preprocessing_id = \"modules/epigenetics_preprocessing.nf\"\n",
    "epi_annotation_script_id = \"bin/epigenomics_annotation.R\"\n",
    "epi_filtering_script_id = \"bin/epigenomics_filtering.R\"\n",
    "epi_imputation_script_id = \"bin/epigenomics_imputation.R\"\n",
    "epi_covariates_correction_script_id = \"bin/CovariateCorrection.R\"\n",
    "epi_subset_features_script_id = \"bin/sort_cols_sd.R\"\n",
    "epi_scaling_script_id = \"bin/epigenomics_scaling.R\"\n",
    "\n",
    "\n",
    "# Define epigenomics pre-processing sub-workflow\n",
    "epi_preprocessing = crate.add(ComputationalWorkflow(crate, epi_preprocessing_id, dest_path=epi_preprocessing_id, properties={\n",
    "    \"@id\": epi_preprocessing_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Epigenetics preprocessing\",\n",
    "    #\"input\": {\"@id\": \"epigenomics_values\"}, #Add later\n",
    "    \"output\": {\"@id\": \"epigenomics_preprocessed_data\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": epi_annotation_script_id},\n",
    "        {\"@id\": epi_filtering_script_id},\n",
    "        {\"@id\": epi_imputation_script_id},\n",
    "        {\"@id\": epi_covariates_correction_script_id},\n",
    "        {\"@id\": epi_subset_features_script_id},\n",
    "        {\"@id\": epi_scaling_script_id}\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_0226\", #annotation\n",
    "                       \"@id\": \"http://purl.obolibrary.org/obo/MS_1001486\", #filtering\n",
    "                       \"@id\": \"http://edamontology.org/operation_3557\", #imputation\n",
    "                       \"@id\": \"http://purl.obolibrary.org/obo/OBI_0200185\", #scaling\n",
    "                       \"@id\": \"http://semanticscience.org/resource/SIO_000594\"} #data transformation\n",
    "}))\n",
    "\n",
    "\n",
    "# Scripts\n",
    "epi_annotation_script = crate.add(File(crate, epi_annotation_script_id, dest_path=epi_annotation_script_id, properties={\n",
    "    \"@id\": epi_annotation_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Epigenomics annotation script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_0226\"} #annotation\n",
    "   }))\n",
    "\n",
    "epi_filtering_script = crate.add(File(crate, epi_filtering_script_id, dest_path=epi_filtering_script_id, properties={\n",
    "    \"@id\": epi_filtering_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Epigenomics filtering script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1001486\"} #filtering\n",
    "}))\n",
    "\n",
    "epi_imputation_script = crate.add(File(crate, epi_imputation_script_id, dest_path=epi_imputation_script_id, properties={\n",
    "    \"@id\": epi_imputation_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Epigenomics imputation script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_3557\"} #imputation\n",
    "}))\n",
    "\n",
    "epi_covariates_correction_script = crate.add(File(crate, epi_covariates_correction_script_id, dest_path=epi_covariates_correction_script_id, properties={\n",
    "    \"@id\": epi_covariates_correction_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Epigenomics covariates correction script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://semanticscience.org/resource/SIO_000594\"} #data transformation #TODO find more specific term\n",
    "}))\n",
    "\n",
    "epi_subset_features_script = crate.add(File(crate, epi_subset_features_script_id, dest_path=epi_subset_features_script_id, properties={\n",
    "    \"@id\": epi_subset_features_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Epigenomics subset features script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://semanticscience.org/resource/SIO_000594\"} #data transformation #TODO find more specific term\n",
    "}))\n",
    "\n",
    "epi_scaling_script = crate.add(File(crate, epi_scaling_script_id, dest_path=epi_scaling_script_id, properties={\n",
    "    \"@id\": epi_scaling_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Epigenomics scaling script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OBI_0200185\"} #scaling\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6ba6b05c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define epigenomics data entities\n",
    "\n",
    "# Define data entity IDs\n",
    "epigenomics_data_id = \"Synthetic_data/synthetic_epigenomics.csv\"\n",
    "epigenomics_meta_id = \"Synthetic_data/synthetic_epigenomics_meta.csv\"\n",
    "\n",
    "epigenomics_data = crate.add(File(crate, epigenomics_data_id, dest_path=epigenomics_data_id, properties={\n",
    "    \"@id\": epigenomics_data_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C153195\"}, #epigenome\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "epigenomics_meta = crate.add(File(crate, epigenomics_meta_id, dest_path=epigenomics_meta_id, properties={\n",
    "    \"@id\": epigenomics_meta_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C52095\"}, #metadata\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "epigenomics_preprocessed = crate.add(DataEntity(crate, \"epigenomics_preprocessed\", properties={\n",
    "    \"@id\": \"epigenomics_preprocessed_data\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C153195\", #epigenome\n",
    "                       \"@id\": \"http://www.ebi.ac.uk/efo/EFO_0004096\" #processed array data file\n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "epi_preprocessing[\"input\"] = [epigenomics_data_id, epigenomics_meta_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3322b5d",
   "metadata": {},
   "source": [
    "## Check if Methylation EPIC file is present, if not download from url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fecfddba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, os.path\n",
    "import zipfile\n",
    "from os import path\n",
    "from io import BytesIO\n",
    "\n",
    "url_methylationEPIC = 'https://webdata.illumina.com/downloads/productfiles/methylationEPIC/infinium-methylationepic-v-1-0-b4-manifest-file-csv.zip'\n",
    "\n",
    "if path.exists(\"EPIC_annotation/raw/MethylationEPIC_v-1-0_B4.csv\") is False:\n",
    "    req = requests.get(url_methylationEPIC)\n",
    "\n",
    "    # Writing the file to the local file system\n",
    "    zipfile= zipfile.ZipFile(BytesIO(req.content))\n",
    "    zipfile.extractall('EPIC_annotation/raw/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b4cb0530",
   "metadata": {},
   "outputs": [],
   "source": [
    "# EPIC annotation files used for epigenomics data\n",
    "\n",
    "# Define data entity IDs\n",
    "EPIC_MOESM1_id = \"EPIC_annotation/raw/13059_2016_1066_MOESM1_ESM.csv\"\n",
    "EPIC_MOESM4_id = \"EPIC_annotation/raw/13059_2016_1066_MOESM4_ESM.csv\"\n",
    "EPIC_MOESM5_id = \"EPIC_annotation/raw/13059_2016_1066_MOESM5_ESM.csv\"\n",
    "Methylation_EPIC_id = \"EPIC_annotation/raw/MethylationEPIC_v-1-0_B4.csv\"\n",
    "Annotation_EPIC_id = \"EPIC_annotation/anno_epic_072017.RData\"\n",
    "\n",
    "\n",
    "Epic_MOESM1 = crate.add(File(crate, EPIC_MOESM1_id, dest_path=EPIC_MOESM1_id, properties={\n",
    "    \"@id\": EPIC_MOESM1_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C43523\" #probe #TODO: find better term(s)\n",
    "                      },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "Epic_MOESM4 = crate.add(File(crate, EPIC_MOESM4_id, dest_path=EPIC_MOESM4_id, properties={\n",
    "    \"@id\": EPIC_MOESM4_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C43523\" #probe #TODO: find better term(s)\n",
    "                      },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "Epic_MOESM5 = crate.add(File(crate, EPIC_MOESM5_id, dest_path=EPIC_MOESM5_id, properties={\n",
    "    \"@id\": EPIC_MOESM5_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C43523\" #probe #TODO: find better term(s)\n",
    "                      },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "Methylation_EPIC = crate.add(File(crate, Methylation_EPIC_id, dest_path=Methylation_EPIC_id, properties={\n",
    "    \"@id\": Methylation_EPIC_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OBI_0002131\" #Illumina Infinium MethylationEPIC BeadChip\n",
    "                            #TODO: find better term(s)\n",
    "                      },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "Annotation_EPIC = crate.add(File(crate, Annotation_EPIC_id, dest_path=Annotation_EPIC_id, properties={\n",
    "    \"@id\": Annotation_EPIC_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"epigenomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": { \"@id\": \"http://edamontology.org/operation_0226\" #annotation\n",
    "                      },\n",
    "    #\"format\":  #TODO add alternative to .RData\n",
    "}))\n",
    "\n",
    "\n",
    "epi_annotation_script[\"hasPart\"] = [EPIC_MOESM1_id, EPIC_MOESM4_id, EPIC_MOESM5_id, Methylation_EPIC_id, Annotation_EPIC_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "428713c4",
   "metadata": {},
   "source": [
    "### Metabolomics pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "622d97ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define data entity IDs\n",
    "mtblmcs_preprocessing_id = \"modules/metabolomics_preprocessing.nf\"\n",
    "mtblmcs_filtering_script_id = \"bin/metabolomics_filter.Rmd\"\n",
    "mtblmcs_normalization_script_id = \"bin/metabolomics_normalization.R\"\n",
    "mtblmcs_scaling_script_id = \"bin/metabolomics_scaling.R\"\n",
    "mtblmcs_concatenate_script_id = \"bin/concatenate_MAF.R\"\n",
    "\n",
    "\n",
    "# Define metabolomics pre-processing sub-workflow\n",
    "mtblmcs_preprocessing = crate.add(ComputationalWorkflow(crate, mtblmcs_preprocessing_id, dest_path=mtblmcs_preprocessing_id, properties={\n",
    "    \"@id\": mtblmcs_preprocessing_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Metabolomics preprocessing\",\n",
    "    #\"input\": #Add later\n",
    "    \"output\": {\"@id\": \"metabolomics_preprocessed_data\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": mtblmcs_filtering_script_id},\n",
    "        {\"@id\": mtblmcs_normalization_script_id},\n",
    "        {\"@id\": mtblmcs_scaling_script_id},\n",
    "        {\"@id\": mtblmcs_concatenate_script_id}\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1001486\", #filtering\n",
    "                       \"@id\": \"http://purl.obolibrary.org/obo/OBI_0200169\" #normalization\n",
    "                        }\n",
    "}))\n",
    "\n",
    "\n",
    "# Define scripts\n",
    "mtblmcs_filtering_script = crate.add(File(crate, mtblmcs_filtering_script_id, dest_path=mtblmcs_filtering_script_id, properties={\n",
    "    \"@id\": mtblmcs_filtering_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Metabolomics filtering script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_4000\"}, #R markdown\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1001486\"} #filtering\n",
    "}))\n",
    "\n",
    "\n",
    "mtblmcs_normalization_script = crate.add(File(crate, mtblmcs_normalization_script_id, dest_path=mtblmcs_normalization_script_id, properties={\n",
    "    \"@id\": mtblmcs_normalization_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Metabolomics normalization script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OBI_0200169\"} #normalization data transformation\n",
    "}))\n",
    "\n",
    "mtblmcs_scaling_script = crate.add(File(crate, mtblmcs_scaling_script_id, dest_path=mtblmcs_scaling_script_id, properties={\n",
    "    \"@id\": mtblmcs_scaling_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Metabolomics scaling script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OBI_0200037\"} #pareto scaling\n",
    "}))\n",
    "\n",
    "mtblmcs_concatenate_script = crate.add(File(crate, mtblmcs_concatenate_script_id, dest_path=mtblmcs_concatenate_script_id, properties={\n",
    "    \"@id\": mtblmcs_concatenate_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Concatenate MAFs script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OBI_0002566\"} #file merge\n",
    "}))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "78cbb03f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define metabolomics data entities\n",
    "\n",
    "metabolomics_data_id = \"Synthetic_data/synthetic_metabolomics.csv\"\n",
    "amines_data_id = \"Synthetic_data/amines_MAF.tsv\"\n",
    "OA_data_id = \"Synthetic_data/OA_MAF.tsv\"\n",
    "steroids_data_id = \"Synthetic_data/steroids_MAF.tsv\" \n",
    "\n",
    "metabolomics_data = crate.add(File(crate, metabolomics_data_id, dest_path=metabolomics_data_id, properties={\n",
    "    \"@id\": metabolomics_data_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"metabolomics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1003084\", #processed data file (not raw)\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/CHEBI_32952\", #amine\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/CHEBI_64709\", #organic acid\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/CHEBI_35341\", #steroid\n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1000914\"} #tsv\n",
    "}))\n",
    "\n",
    "amines_data = crate.add(File(crate, amines_data_id, dest_path=amines_data_id, properties={\n",
    "    \"@id\": amines_data_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"metabolomics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1003084\", #processed data file (not raw)\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/CHEBI_32952\", #amine\n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1000914\"} #tsv\n",
    "}))\n",
    "\n",
    "OA_data = crate.add(File(crate, OA_data_id, dest_path=OA_data_id, properties={\n",
    "    \"@id\": OA_data_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"metabolomics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1003084\", #processed data file (not raw)\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/CHEBI_64709\", #organic acid\n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1000914\"} #tsv\n",
    "}))\n",
    "\n",
    "steroids_data = crate.add(File(crate, steroids_data_id, dest_path=steroids_data_id, properties={\n",
    "    \"@id\": steroids_data_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"metabolomics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1003084\", #processed data file (not raw)\n",
    "                        \"@id\": \"http://purl.obolibrary.org/obo/CHEBI_35341\", #steroid\n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1000914\"} #tsv\n",
    "}))\n",
    "\n",
    "metabolomics_preprocessed = DataEntity(crate, \"metabolomics_preprocessed\", properties={\n",
    "    \"@id\": \"metabolomics_preprocessed_data\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"metabolomics_preprocessed\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1003084\", #processed data file (not raw),\n",
    "                       \"@id\": \"http://purl.obolibrary.org/obo/OBI_0000451\" #normalized data set\n",
    "                        },\n",
    "    \"format\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1000914\"} #tsv\n",
    "})\n",
    "\n",
    "mtblmcs_preprocessing[\"input\"] = [metabolomics_data_id, amines_data_id, OA_data_id, steroids_data_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0908918",
   "metadata": {},
   "source": [
    "### Phenotypes preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "cc279e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define data entity IDs\n",
    "cbcl_imputation_mca_id = \"modules/CBCL_MCA.nf\"\n",
    "cbcl_imputation_mca_script_id = \"bin/CBCL_filter_impute_MCA.Rmd\"\n",
    "\n",
    "\n",
    "# Define behavioral data pre-processing sub-workflow\n",
    "cbcl_imputation_mca = crate.add(ComputationalWorkflow(crate, cbcl_imputation_mca_id, dest_path=cbcl_imputation_mca_id, properties={\n",
    "    \"@id\": cbcl_imputation_mca_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"CBCL imputation\",\n",
    "    \"input\": {\"@id\": \"behavioral_data\"},\n",
    "    \"output\": {\"@id\": \"behavioral_data\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": cbcl_imputation_mca_script_id }\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_3557\", #imputation\n",
    "                       \"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning \n",
    "                       }\n",
    "}))\n",
    "\n",
    "\n",
    "# Define script\n",
    "cbcl_imputation_mca_script = crate.add(File(crate, cbcl_imputation_mca_script_id, dest_path=cbcl_imputation_mca_script_id , properties={\n",
    "    \"@id\": cbcl_imputation_mca_script_id ,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"Filter, impute CBCL data and MCA\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_4000\"}, #R markdown\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_3557\", #imputation\n",
    "                       \"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning \n",
    "                       }\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "fcb5eb2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define phenotypic data entities\n",
    "\n",
    "phenotype_covariates_id = \"Synthetic_data/synthetic_phenotype_covariates_data.csv\"\n",
    "behavioral_data_id = \"Synthetic_data/synthetic_cbcl_data.csv\"\n",
    "\n",
    "phenotype_covariates = crate.add(File(crate, phenotype_covariates_id, dest_path=phenotype_covariates_id, properties={\n",
    "    \"@id\": phenotype_covariates_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"phenotype_covariates\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": { \"@id\": \"http://purl.obolibrary.org/obo/NCIT_C16977\" #Phenotype\n",
    "                        },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "behavioral_data = crate.add(File(crate, behavioral_data_id, dest_path=behavioral_data_id, properties={\n",
    "    \"@id\": behavioral_data_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"behavioral_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": { \"@id\": \"http://www.ebi.ac.uk/efo/EFO_0005661\" #CBCL assessment\n",
    "                        },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e19b31c",
   "metadata": {},
   "source": [
    "### Map omics files \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b08d529a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define data entity IDs\n",
    "id_mapping_id = \"modules/map_IDs.nf\"\n",
    "id_file_id = \"Synthetic_data/ACTIONdemonstrator_XOmics_IDs_synthetic.csv\"\n",
    "id_mapping_script_id = \"bin/map_IDs.py\"\n",
    "\n",
    "id_file = crate.add(File(crate, id_file_id, dest_path=id_file_id, properties={\n",
    "    \"@id\": id_file_id,\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"behavioral_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": { \"@id\": \"http://purl.enanomapper.org/onto/ENM_9000071\" #sample identifier\n",
    "                        },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "}))\n",
    "\n",
    "# Define module of sample identifier mapping\n",
    "id_mapping = crate.add(ComputationalWorkflow(crate, id_mapping_id, dest_path=id_mapping_id, properties={\n",
    "    \"@id\": id_mapping_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Sample ID mapping\",\n",
    "    \"input\": {\"@id\": \"metabolomics_preprocessed_data\",\n",
    "              \"@id\": \"epigenomics_preprocessed_data\",\n",
    "              \"@id\": id_file_id},\n",
    "    \"output\": {\"@id\": \"metabolomics_preprocessed_data\",\n",
    "              \"@id\": \"epigenomics_preprocessed_data\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": id_mapping_script_id}\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_3282\"} #ID mapping\n",
    "}))\n",
    "\n",
    "\n",
    "# ID mapping script\n",
    "id_mapping_script = crate.add(File(crate, id_mapping_script_id, dest_path=id_mapping_script_id, properties={\n",
    "    \"@id\": id_mapping_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"ID mapping script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3996\"}, #Python script\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_3282\"} #ID mapping\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac3835cc",
   "metadata": {},
   "source": [
    "### Principal Component Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9c63a3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define data entity IDs\n",
    "pca_id = \"modules/pca.nf\"\n",
    "pca_script_id = \"bin/pca.R\"\n",
    "\n",
    "\n",
    "# Define pca sub-workflow\n",
    "pca = crate.add(ComputationalWorkflow(crate, pca_id, dest_path=pca_id, properties={\n",
    "    \"@id\": pca_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Principal Component Analysis\",\n",
    "    \"input\": {\"@id\": \"processed_omics_data\"},\n",
    "    \"output\": {\"@id\": \"pca_report\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": pca_script_id}\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C49291\" #PCA\n",
    "                        }\n",
    "}))\n",
    "\n",
    "\n",
    "# PCA script\n",
    "pca_script = crate.add(File(crate, pca_script_id, dest_path=pca_script_id, properties={\n",
    "    \"@id\": pca_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"PCA script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/NCIT_C49291\"} #PCA\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "418c2e07",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define PCA input/output entities\n",
    "\n",
    "processed_omics_data = DataEntity(crate, \"omics_data.csv\", properties={\n",
    "    \"@id\": \"processed_omics_data\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"processed_omics_data\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/MS_1003084\", #processed data file \n",
    "                     \"@id\": \"http://edamontology.org/topic_3391\" #omics\n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3752\"} #csv\n",
    "})\n",
    "\n",
    "#TODO What should be output here? .pdf file of plots?\n",
    "\n",
    "pca_report = DataEntity(crate, \"pca.pdf\", properties={\n",
    "    \"@id\": \"pca_report\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"pca_report\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/data_2884\" #plot (multiple) \n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3508\"} #pdf\n",
    "})\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fec0eeb3",
   "metadata": {},
   "source": [
    "### Similarity Network Fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "98c39c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define data entity IDs\n",
    "snf_id = \"modules/snf.nf\"\n",
    "snf_script_id = \"bin/perform_snf.py\"\n",
    "snf_analysis_script_id = \"bin/snf_analysis.ipynb\"\n",
    "snf_gee_script_id = \"bin/snf_gee_analysis.ipynb\"\n",
    "\n",
    "\n",
    "# Define SNF sub-workflow\n",
    "snf = crate.add(ComputationalWorkflow(crate, snf_id, dest_path=snf_id, properties={\n",
    "    \"@id\": snf_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Similarity Network Fusion\",\n",
    "    \"input\": {\"@id\": \"processed_omics_data\"},\n",
    "    \"output\": {\"@id\": \"snf_report\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": snf_script_id},\n",
    "        {\"@id\": snf_analysis_script_id},\n",
    "        {\"@id\": snf_gee_script_id}\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://edamontology.org/operation_3432\", #Clustering \n",
    "                       \"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                         }\n",
    "}))\n",
    "\n",
    "\n",
    "# SNF scripts\n",
    "snf_script = crate.add(File(crate, snf_script_id, dest_path=snf_script_id, properties={\n",
    "    \"@id\": snf_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"SNF script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3996\"}, #Python script\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                       \"@id\": \"http://edamontology.org/operation_3432\", #Clustering \n",
    "                         }\n",
    "}))\n",
    "\n",
    "snf_analysis_script = crate.add(File(crate, snf_analysis_script_id, dest_path=snf_analysis_script_id, properties={\n",
    "    \"@id\": snf_analysis_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"SNF downstream analysis script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3996\"}, #Python script\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                       \"@id\": \"http://edamontology.org/operation_3432\",  #Clustering \n",
    "                       \"@id\": \"http://semanticscience.org/resource/SIO_000449\" #plot\n",
    "                         }\n",
    "}))\n",
    "\n",
    "snf_gee_script = crate.add(File(crate, snf_gee_script_id, dest_path=snf_gee_script_id, properties={\n",
    "    \"@id\": snf_gee_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"SNF GEE models\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3996\"}, #Python script\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\" #Unsupervised learning\n",
    "                         }\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0becffc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define SNF input/output entities\n",
    "\n",
    "snf_report = crate.add(DataEntity(crate, \"snf_report\", properties={\n",
    "    \"@id\": \"snf_report\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"snf_report\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://semanticscience.org/resource/SIO_000449\" #plot \n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3508\"} #pdf\n",
    "}))\n",
    "\n",
    "#TODO add snf matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29be1ef2",
   "metadata": {},
   "source": [
    "### Multi-Omics Factor Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "157b0838",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define data entity IDs\n",
    "mofa_id = \"modules/mofa.nf\"\n",
    "mofa_script_id = \"bin/mofa.R\"\n",
    "mofa_analysis_script_id = \"bin/MOFA_downstream_analysis_report.Rmd\"\n",
    "mofa_gee_script_id = \"bin/MOFA_downstream_analysis_report_gee.Rmd\"\n",
    "\n",
    "\n",
    "# Define MOFA module\n",
    "mofa = crate.add(ComputationalWorkflow(crate, mofa_id, dest_path=mofa_id, properties={\n",
    "    \"@id\": mofa_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\", \"ComputationalWorkflow\"],\n",
    "    \"name\": \"Multi-Omics Factor Analysis\",\n",
    "    \"programmingLanguage\": {\"@id\": \"https://w3id.org/workflowhub/workflow-ro-crate#nextflow\"},\n",
    "    \"input\": {\"@id\": \"processed_omics_data\"},\n",
    "    \"output\": {\"@id\": \"mofa_model\"},\n",
    "    \"hasPart\": [\n",
    "        {\"@id\": mofa_script_id},\n",
    "        {\"@id\": mofa_analysis_script_id},\n",
    "        {\"@id\": mofa_gee_script_id}\n",
    "    ],\n",
    "    \"license\": \"https://opensource.org/licenses/MIT\",\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                       \"@id\":  \"http://edamontology.org/topic_3474\"   #Machine learning\n",
    "                         }\n",
    "}))\n",
    "\n",
    "\n",
    "# MOFA script\n",
    "mofa_script = crate.add(File(crate, mofa_script_id, dest_path=mofa_script_id, properties={\n",
    "    \"@id\": mofa_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"MOFA script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_3999\"}, #Rscript\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                       \"@id\":  \"http://edamontology.org/topic_3474\"   #Machine learning\n",
    "                         }\n",
    "}))\n",
    "\n",
    "# MOFA script\n",
    "mofa_analysis_script = crate.add(File(crate, mofa_analysis_script_id, dest_path=mofa_analysis_script_id, properties={\n",
    "    \"@id\": mofa_analysis_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"MOFA downstream analysis script\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_4000\"}, #R markdown\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                       \"@id\":  \"http://edamontology.org/topic_3474\"   #Machine learning\n",
    "                         }\n",
    "}))\n",
    "\n",
    "# MOFA GEE script\n",
    "mofa_analysis_gee_script = crate.add(File(crate, mofa_gee_script_id, dest_path=mofa_gee_script_id, properties={\n",
    "    \"@id\": mofa_gee_script_id,\n",
    "    \"@type\": [\"File\", \"SoftwareSourceCode\"],\n",
    "    \"name\": \"MOFA downstream analysis script with GEE\",\n",
    "    \"programmingLanguage\": {\"@id\": \"http://edamontology.org/format_4000\"}, #R markdown\n",
    "    \"additionalType\": {\"@id\": \"http://purl.enanomapper.org/onto/ENM_8000003\", #Unsupervised learning\n",
    "                       \"@id\":  \"http://edamontology.org/topic_3474\"   #Machine learning\n",
    "                         }\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c984bb2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define MOFA output mode\n",
    "\n",
    "mofa_model = crate.add(DataEntity(crate, \"mofa_model\", properties={\n",
    "    \"@id\": \"mofa_model\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"mofa_model\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/STATO_0000107\" #statistical model \n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3590\"} #HDF5\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31de70be",
   "metadata": {},
   "source": [
    "## Data entities on Root directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "9faca933",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rocrate.model.data_entity import DataEntity\n",
    "\n",
    "nextflow_config = crate.add_file( \"nextflow.config\", properties={\n",
    "    \"@id\": \"nextflow.config\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Nextflow configuration file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/ONTOAVIDA_00000001\" #configuration file \n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3464\"} #JSON\n",
    "})\n",
    "\n",
    "dre_config = crate.add_file( \"dre.config\", properties={\n",
    "    \"@id\": \"dre.config\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Nextflow configuration file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/ONTOAVIDA_00000001\" #configuration file \n",
    "                         },\n",
    "    \"format\": {\"@id\": \"http://edamontology.org/format_3464\"} #JSON\n",
    "})\n",
    "\n",
    "readme = crate.add_file(\"README.md\", properties={\n",
    "    \"@id\": \"README.md\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Nextflow configuration file\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OMIT_00055391\" #Documentation\n",
    "                         },\n",
    "    #\"format\": {} #TODO: markdown ontology term needed here. \n",
    "})\n",
    "\n",
    "action_documentation = crate.add_file(\"Documentation.md\", properties={\n",
    "    \"@id\": \"ACTION_documentation.md\",\n",
    "    \"@type\": \"FormalParameter\",\n",
    "    \"name\": \"Documentation on Workflow\",\n",
    "    #\"valueRequired\": true,\n",
    "    \"additionalType\": {\"@id\": \"http://purl.obolibrary.org/obo/OMIT_00055391\" #Documentation\n",
    "                         },\n",
    "    #\"format\": {} #TODO: markdown ontology term needed here. \n",
    "})\n",
    "\n",
    " # Define the diagram that provides an overview of the main workflow\n",
    "\n",
    "diagram = crate.add_file(\"flowchart.png\", properties={\n",
    "    \"@id\": \"flowchart.png\",\n",
    "    \"@type\": [\"File\", \"ImageObject\"],\n",
    "    \"name\": \"Workflow overview\" ,\n",
    "    \"about\": {\"@id\": \"action.nf\"}\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18ebd945",
   "metadata": {},
   "source": [
    "## Add entities to the main workflow entity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "03869126",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add entities/attributes to the workflow\n",
    "\n",
    "MainWorkflow[\"author\"] = [Anna_Niehues, Casper_de_Visser, Fiona_Hagenbeek, Naama_Karu, Alida_Kindt, Purva_Kulkarni, Rene_Pool, Dorret_Boomsma, Jenny_van_Dongen, Alain_van_Gool, PeterBram_t_Hoen]\n",
    "MainWorkflow[\"programmingLanguage\"] =  Nextflow\n",
    "MainWorkflow[\"image\"] = diagram\n",
    "MainWorkflow[\"config\"] = [nextflow_config, dre_config]\n",
    "MainWorkflow[\"sdPublisher\"] =  x_omics\n",
    "MainWorkflow[\"hasPart\"] = [missing_data_heatmap, epi_preprocessing, mtblmcs_preprocessing, cbcl_imputation_mca, id_mapping, pca, snf, mofa]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b2db106",
   "metadata": {},
   "source": [
    "## Add publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "1fa74fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rio Journal abstract\n",
    "\n",
    "RIO_abstract = crate.add(ContextEntity(crate, \"https://doi.org/10.3897/rio.8.e94042\", properties={\n",
    "    \"@id\": \"https://doi.org/10.3897/rio.8.e94042\",\n",
    "    \"@type\": [\"ScholartlyArtcile\", \"CreativeWork\"],\n",
    "    \"name\": \"A Multi-omics Data Analysis Workflow Packaged as a FAIR Digital Object\",\n",
    "    \"dateCreated\": \"25-08-2022\",\n",
    "    \"keywords\": [\"Multi-omics\", \"Metabolomics\", \"Epigenomics\", \"Behavioral data\", \"FAIR\"],\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "796a493b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add information / link entitites to the crate\n",
    "\n",
    "crate.mainEntity = MainWorkflow\n",
    "crate.name =  \"X-omics ACTIONdemonstrator analysis workflow\"\n",
    "crate.author = [Anna_Niehues, Casper_de_Visser, Fiona_Hagenbeek, Naama_Karu, Alida_Kindt, Purva_Kulkarni, Rene_Pool, Dorret_Boomsma, Jenny_van_Dongen, Alain_van_Gool, PeterBram_t_Hoen]\n",
    "crate.license = \"https://opensource.org/licenses/MIT\"\n",
    "crate.keywords =  [\"Multi-omics\", \"Metabolomics\", \"Epigenomics\", \"Behavioral data\", \"FAIR\"]\n",
    "crate.datePublished = str(today)\n",
    "crate.description = \"This workflow is designed to analyze to a multi-omics data set that comprises genome-wide DNA methylation profiles, targeted metabolomics, and behavioral data of two cohorts that participated in the ACTION Biomarker Study (ACTION, Aggression in Children: Unraveling gene-environment interplay to inform Treatment and InterventiON strategies. (Boomsma 2015, Bartels 2018, Hagenbeek 2020, van Dongen 2021, Hagenbeek 2022). The ACTION-NTR cohort consists of twins that are either longitudinally concordant or discordant for childhood aggression. The ACTION-Curium-LUMC cohort consists of children referred to the Dutch LUMC Curium academic center for child and youth psychiatry. With the joint analysis of multi-omics data and behavioral data, we aim to identify substructures in the ACTION-NTR cohort and link them to aggressive behavior. First, the individuals are clustered using Similarity Network Fusion (SNF, Wang 2014), and latent feature dimensions are uncovered using different unsupervised methods including Multi-Omics Factor Analysis (MOFA) (Argelaguet 2018) and Multiple Correspondence Analysis (MCA, L√™ 2008, Husson 2017). In a second step, we determine correlations between -omics and phenotype dimensions, and use them to explain the subgroups of individuals from the ACTION-NTR cohort. In order to validate the results, we project data of the ACTION-Curium-LUMC cohort onto the latent dimensions and determine if correlations between omics and phenotype data can be reproduced.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "96b0d8ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to JSON-LD\n",
    "\n",
    "#crate.write(\"exp_crate\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "ea56696d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ro-crate_container/volume/ACTIONdemonstrator_workflow/ro-crate/exp_crate.zip'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save to ZIP-file\n",
    "\n",
    "crate.write_zip(\"ro-crate/exp_crate.zip\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "vscode": {
   "interpreter": {
    "hash": "f8a604dd50995c9592d795865f70db7e3507fb707354087e2ac278c961b96283"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
